backend/clickhouse/traces.go

```diff
@@ -5,58 +5,86 @@ import (
+	EventsAttributes []map[string]string `db:"Events.Attributes"`
+	LinksTraceId     []string            `db:"Links.TraceId"`
+	LinksSpanId      []string            `db:"Links.SpanId"`
+	LinksTraceState  []string            `db:"Links.TraceState"`
+	LinksAttributes  []map[string]string `db:"Links.Attributes"`
+}
+
 func (client *Client) BatchWriteTraceRows(ctx context.Context, traceRows []*TraceRow) error {
 	if len(traceRows) == 0 {
 		return nil
 	}
 
-	batch, err := client.conn.PrepareBatch(ctx, fmt.Sprintf("INSERT INTO %s", TracesTable))
-	if err != nil {
-		return e.Wrap(err, "could not prepare traces batch")
-	}
-
-	for _, traceRow := range traceRows {
-		// Was not able to figure out a way to insert nested values using a struct +
-		// AppendStruct, so had to pull them out to individual arrays.
+	rows := lo.Map(traceRows, func(traceRow *TraceRow, _ int) interface{} {
 		traceTimes, traceNames, traceAttrs := convertEvents(traceRow)
 		linkTraceIds, linkSpanIds, linkStates, linkAttrs := convertLinks(traceRow)
 
-		err = batch.Append(
-			traceRow.Timestamp,
-			traceRow.UUID,
-			traceRow.TraceId,
-			traceRow.SpanId,
-			traceRow.ParentSpanId,
-			traceRow.ProjectId,
-			traceRow.SecureSessionId,
-			traceRow.TraceState,
-			traceRow.SpanName,
-			traceRow.SpanKind,
-			traceRow.Duration,
-			traceRow.ServiceName,
-			traceRow.ServiceVersion,
-			traceRow.TraceAttributes,
-			traceRow.StatusCode,
-			traceRow.StatusMessage,
-			traceTimes,
-			traceNames,
-			traceAttrs,
-			linkTraceIds,
-			linkSpanIds,
-			linkStates,
-			linkAttrs,
-		)
-		if err != nil {
-			return err
+		row := ClickhouseTraceRow{
+			Timestamp:        traceRow.Timestamp,
+			UUID:             traceRow.UUID,
+			TraceId:          traceRow.TraceId,
+			SpanId:           traceRow.SpanId,
+			ParentSpanId:     traceRow.ParentSpanId,
+			ProjectId:        traceRow.ProjectId,
+			SecureSessionId:  traceRow.SecureSessionId,
+			TraceState:       traceRow.TraceState,
+			SpanName:         traceRow.SpanName,
+			SpanKind:         traceRow.SpanKind,
+			Duration:         traceRow.Duration,
+			ServiceName:      traceRow.ServiceName,
+			ServiceVersion:   traceRow.ServiceVersion,
+			TraceAttributes:  traceRow.TraceAttributes,
+			StatusCode:       traceRow.StatusCode,
+			StatusMessage:    traceRow.StatusMessage,
+			EventsTimestamp:  traceTimes,
+			EventsName:       traceNames,
+			EventsAttributes: traceAttrs,
+			LinksTraceId:     linkTraceIds,
+			LinksSpanId:      linkSpanIds,
+			LinksTraceState:  linkStates,
+			LinksAttributes:  linkAttrs,
 		}
-	}
 
-	return batch.Send()
+		return row
+	})
+
+	ib := sqlbuilder.NewStruct(new(ClickhouseTraceRow)).InsertInto(TracesTable, rows...)
+	sql, args := ib.BuildWithFlavor(sqlbuilder.ClickHouse)
+
+	chCtx := clickhouse.Context(ctx, clickhouse.WithSettings(clickhouse.Settings{
+		"async_insert":          1,
+		"wait_for_async_insert": 1,
+	}))
+	fmt.Printf("::: sql (traces): %s\n", sql)
+	return client.conn.Exec(chCtx, sql, args...)
```

Comment:
@mayberryzane I keep hitting this error trying to write traces to Clickhouse:

```sh
ERRO[0112]/Users/chris/code/highlight/backend/worker/kafka_worker.go:293 github.com/highlight-run/highlight/backend/worker.(*KafkaBatchWorker).flushTraces() failed to batch write traces to clickhouse    error="code: 62, message: Cannot parse expression of type Array(DateTime) here: , , , , , , ), (toDateTime('2023-08-09 03:14:17'), '454360f2-6eef-406e-9ec6-01d4c0ca422c', 'b2e3dd9d868e30cc962b5d51b6c8fa91', 'e500d66b9c2a6de1', '', 1, '', '': While executing WaitForAsyncInsert"
```

I thought maybe it was writing to `Events.Timestamp`, but I tried converting that to a string even and I'm still getting that error. Do you have any ideas what might be going on here?

Reply:
I ran into the same thing when cutting over sessions to use async inserts, I had to change any arrays in my model to have `clickhouse.ArraySet` type:
https://sourcegraph.com/github.com/highlight/highlight/-/blob/backend/clickhouse/sessions.go?L20

Reply:
Ahh, nice - thank you!